commit 0f22072ab50cac7983f9660d33974b45184da4f9
Author: Dan Rosenberg <drosenberg@vsecurity.com>
Date:   Fri Apr 29 15:48:07 2011 +0100

    ARM: 6891/1: prevent heap corruption in OABI semtimedop
    
    When CONFIG_OABI_COMPAT is set, the wrapper for semtimedop does not
    bound the nsops argument.  A sufficiently large value will cause an
    integer overflow in allocation size, followed by copying too much data
    into the allocated buffer.  Fix this by restricting nsops to SEMOPM.
    Untested.
    
    Cc: stable@kernel.org
    Signed-off-by: Dan Rosenberg <drosenberg@vsecurity.com>
    Signed-off-by: Russell King <rmk+kernel@arm.linux.org.uk>

diff --git a/arch/arm/kernel/sys_oabi-compat.c b/arch/arm/kernel/sys_oabi-compat.c
index 4ad8da1..af0aaeb 100644
--- a/arch/arm/kernel/sys_oabi-compat.c
+++ b/arch/arm/kernel/sys_oabi-compat.c
@@ -311,7 +311,7 @@ asmlinkage long sys_oabi_semtimedop(int semid,
 	long err;
 	int i;
 
-	if (nsops < 1)
+	if (nsops < 1 || nsops > SEMOPM)
 		return -EINVAL;
 	sops = kmalloc(sizeof(*sops) * nsops, GFP_KERNEL);
 	if (!sops)
